src/can.rs

```diff
@@ -1,89 +1,141 @@
-pins! {
-    CAN1 => (PA12, PA11),
-    CAN1 => (PB9, PB8),
-    CAN1 => (PD1, PD0),
-    CAN1 => (PH13, PI9),
-    CAN2 => (PB13, PB12),
-    CAN2 => (PB6, PB5),
+#[cfg(any(feature = "stm32f412", feature = "stm32f413", feature = "stm32f423"))]
+mod pb9_pb8_af8 {
+    use crate::gpio::{
+        gpiob::{PB8, PB9},
+        AF8,
+    };
+    use crate::pac::CAN1;
+    pins! { CAN1 => (PB9<AF8>, PB8<AF8>), }
+}
+
+#[cfg(any(
+    feature = "stm32f405",
+    feature = "stm32f407",
+    feature = "stm32f415",
+    feature = "stm32f417",
+    feature = "stm32f427",
+    feature = "stm32f429",
+    feature = "stm32f437",
+    feature = "stm32f439",
+    feature = "stm32f446",
+    feature = "stm32f469",
+    feature = "stm32f479"
+))]
+mod pb9_pb8_af9 {
+    use crate::gpio::{
+        gpiob::{PB8, PB9},
+        AF9,
+    };
+    use crate::pac::CAN1;
+    pins! { CAN1 => (PB9<AF9>, PB8<AF9>), }
+}
+
+#[cfg(any(feature = "stm32f412", feature = "stm32f413", feature = "stm32f423"))]
+mod pg1_pg0 {
+    use crate::gpio::{
+        gpiog::{PG0, PG1},
+        AF9,
+    };
+    use crate::pac::CAN1;
+    pins! { CAN1 => (PG1<AF9>, PG0<AF9>), }
 }
 
-/*
-    See DS10693 Rev 9 Table 11.
-*/
-#[cfg(feature = "stm32f446")]
-pins! {
-    CAN1 => (PA12, PA11),
-    CAN1 => (PB9, PB8),
-    CAN1 => (PD1, PD0),
-    CAN2 => (PB13, PB12),
-    CAN2 => (PB6, PB5),
+#[cfg(any(feature = "stm32f412", feature = "stm32f413", feature = "stm32f423"))]
+mod pg12_pg11 {
+    use crate::gpio::{
+        gpiog::{PG11, PG12},
+        AF9,
+    };
+    use crate::pac::CAN2;
+    pins! { CAN2 => (PG12<AF9>, PG11<AF9>), }
+}
+
+#[cfg(any(
+    feature = "stm32f405",
+    feature = "stm32f407",
+    feature = "stm32f415",
+    feature = "stm32f417",
+    feature = "stm32f427",
+    feature = "stm32f429",
+    feature = "stm32f437",
+    feature = "stm32f439",
+    feature = "stm32f469",
+    feature = "stm32f479"
+))]
+mod ph13_pi9 {
+    use crate::gpio::{gpioh::PH13, gpioi::PI9, AF9};
+    use crate::pac::CAN1;
+    pins! { CAN1 => (PH13<AF9>, PI9<AF9>), }
 }
 
 /// Enable/disable peripheral
 pub trait Enable: sealed::Sealed {
+    /// Enables this peripheral by setting the associated enable bit in an RCC enable register
     fn enable();
 }
 
+/// Implements sealed::Sealed and Enable for a CAN peripheral (e.g. CAN1)
+///
+/// $peren is the index in RCC_APB1ENR of the enable bit for the CAN peripheral.
 macro_rules! bus {
     ($($PER:ident => ($peren:literal),)+) => {
         $(
-            impl sealed::Sealed for crate::pac::$PER {}
-            impl Enable for crate::pac::$PER {
+            impl crate::can::sealed::Sealed for crate::pac::$PER {}
+            impl crate::can::Enable for crate::pac::$PER {
                 #[inline(always)]
                 fn enable() {
                     unsafe {
-                        let rcc = &(*RCC::ptr());
-                        bb::set(&rcc.apb1enr, $peren)
+                        let rcc = &(*crate::pac::RCC::ptr());
```

Comment:
Doesn't the peripheral need to be reset too?

Reply:
I've seen some drivers that reset their peripherals and some that don't. This is the same as how the code worked before I changed it.

Anyway, I do think that resetting the peripheral is a good idea to avoid misconfiguration problems. I'll add a commit for that.